tao_compiler/mlir/disc/transforms/shape_utils.h

```diff
@@ -0,0 +1,336 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "llvm/ADT/EquivalenceClasses.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+
+#ifndef TENSORFLOW_COMPILER_MLIR_XLA_SHAPE_UTILS_H_
+#define TENSORFLOW_COMPILER_MLIR_XLA_SHAPE_UTILS_H_
+
+namespace mlir {
+namespace disc_ral {
+
+// Represents a symbolic dimension.
+class SymbolDim {
+ public:
+  SymbolDim(int64_t dim_size = ShapedType::kDynamicSize) : dimSize_(dim_size) {}
+
+  int64_t getDimSize() { return dimSize_; }
+
+  bool isDynamic() { return getDimSize() == ShapedType::kDynamicSize; }
+
+  LogicalResult Merge(SymbolDim* other);
+
+ private:
+  int64_t dimSize_;
+};
+
+// Represents a symbolic ranked shape.
+class SymbolShape {
+ public:
+  explicit SymbolShape(SmallVector<SymbolDim*, 4> dims = {})
+      : dims_(std::move(dims)) {}
+
+  int rank() { return dims_.size(); }
+
+  void setSymbolDims(SmallVector<SymbolDim*, 4> dims) {
+    dims_ = std::move(dims);
+  }
+
+  ArrayRef<SymbolDim*> getSymbolDims() { return dims_; }
+
+  SmallVector<int64_t, 4> getDims() {
+    SmallVector<int64_t, 4> dims;
+    for (SymbolDim* dim : dims_) dims.push_back(dim->getDimSize());
+    return dims;
+  }
+
+  SymbolDim* getSymbolDim(int64_t dim) {
+    assert(dim < dims_.size());
+    return dims_[dim];
+  }
+
+  void setSymbolDim(int64_t dim, SymbolDim* symbolDim) {
+    assert(dim < dims_.size());
+    dims_[dim] = symbolDim;
+  }
+
+  bool operator==(const SymbolShape& other) { return other.dims_ == dims_; }
+
+ private:
+  SmallVector<SymbolDim*, 4> dims_;
+};
+
+using llvm::EquivalenceClasses;
+class ValueWrapper {
+ public:
+  explicit ValueWrapper(Value value) : value_(std::move(value)) {}
+
+  Value getValue() const { return value_; }
+
+  bool operator==(const ValueWrapper& rhs) const {
+    return getValue() == rhs.getValue();
+  }
+
+ private:
+  Value value_;
+};
+
+bool operator<(const ValueWrapper& lhs, const ValueWrapper& rhs);
+
+// Shape analysis for propagating and analyzing known shape information in
+// compilation time in a given operation.
+class ShapeAnalysis {
+ public:
+  explicit ShapeAnalysis(Operation* op) : op_(op) {}
+
+  struct DimValue {
+    enum State : int32_t { FOLD = 0, UNFOLD = 1, INVALID = 2 } state;
+    int64_t foldVal = INT64_MIN;
+    Value unfoldVal = nullptr;
+    DimValue() { state = INVALID; }
+    explicit DimValue(Value val) {
+      unfoldVal = val;
+      state = (val == nullptr) ? INVALID : UNFOLD;
+    }
+    explicit DimValue(int64_t val) {
+      foldVal = val;
+      state = (val >= mlir::ShapedType::kDynamicSize) ? FOLD : INVALID;
```

Comment:
What's the meaning if `val == mlir::ShapedType::kDynamicSize`?

Reply:
This should be buggy. Fixed. Thanks!